File: frontend/src/Services/LecturerProfileService.ts
import axios from "axios";
import { LecturerProfileModel } from "../Interfaces/LecturerProfileModel";

export class LecturerProfileService {
    private static readonly baseURL = "https://localhost:7085/api/LecturerProfile";

    private static getAuthHeaders() {
        const token = localStorage.getItem("token");
        return {
            Authorization: token ? `Bearer ${token}` : "",
        };
    }

    public static async getLecturerProfileAsync(): Promise<LecturerProfileModel> {
        try {
            const response = await axios.get(this.baseURL, {
                headers: this.getAuthHeaders(),
            });
            return response.data;
        } catch (error) {
            console.error("Error retrieving the lecturer profile:", error);
            throw error;
        }
    }

    public static async updateLecturerProfileAsync(model: LecturerProfileModel): Promise<LecturerProfileModel> {
        try {
            const response = await axios.put(`${this.baseURL}/${model.id}`, model, {
                headers: this.getAuthHeaders(),
            });
            return response.data;
        } catch (error) {
            console.error("Error updating lecturer profile:", error);
            throw error;
        }
    }
}

File: Auto-Schedule/Application/Services/LecturerProfileService.cs
using AutoMapper;
using Domain.Interface;
using Domain.Model;
using Infrastructure.Data;
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;

namespace Application.Services
{
    public class LecturerProfileService : ILecturerProfileService
    {
        private readonly AppDbContext _context;
        private readonly IMapper _mapper;

        public LecturerProfileService(AppDbContext context, IMapper mapper)
        {
            _context = context;
            _mapper = mapper;
        }

        public async Task<LecturerProfileModel> GetLecturerProfileAsync(Guid id, CancellationToken cancellationToken)
        {
            try
            {
                var lecturer = await _context.Lectures.FirstOrDefaultAsync(l => l.Id == id, cancellationToken);
                if (lecturer == null)
                {
                    throw new KeyNotFoundException("Lecturer not found.");
                }
                return _mapper.Map<LecturerProfileModel>(lecturer);
            }
            catch (OperationCanceledException ex)
            {
                throw new OperationCanceledException("The operation was cancelled.", ex);
            }
            catch (Exception ex)
            {
                throw new Exception("An error occurred while retrieving the lecturer profile.", ex);
            }
        }

        public async Task<LecturerProfileModel> UpdateLecturerProfileAsync(LecturerProfileModel lecturerProfile, CancellationToken cancellationToken)
        {
            try
            {
                var lecturer = await _context.Lectures.FindAsync(new object[] { lecturerProfile.Id }, cancellationToken);
                if (lecturer == null)
                {
                    throw new KeyNotFoundException("Lecturer not found!");
                }

                _mapper.Map(lecturerProfile, lecturer);
                await _context.SaveChangesAsync(cancellationToken);
                return lecturerProfile;
            }
            catch (DbUpdateConcurrencyException ex)
            {
                throw new DbUpdateConcurrencyException("A concurrency error occurred while updating the profile.", ex);
            }
            catch (DbUpdateException ex)
            {
                throw new DbUpdateException("A database error occurred while updating the profile.", ex);
            }
            catch (Exception ex)
            {
                throw new Exception("An error occurred while updating the profile.", ex);
            }
        }
    }
}
